### ISSUE TITLE
System.Xml.XmlDocument.Save() unexpectedly creates UTF-8 files *with BOM*

### ISSUE THREAD

**Issue by mklement0** [dotnet/runtime#28218]
`System.Xml.XmlDocument.Save()`, when given a file path to write to, unexpectedly creates UTF-8 encoded files _with BOM_ if the document's XML declaration has an `encoding="UTF-8"` attribute.

(By contrast, the absence of the `encoding` attribute or absence of an XML declaration altogether causes a BOM-_less_ UTF-8 file to be created, as expected.)

Repro in PowerShell Core (on any supported platform; PowerShell's `[xml]` cast instantiates a `System.Xml.XmlDocument` from the specified XML string):

```powershell
([xml] '<?xml version="1.0" encoding="UTF-8" ?><foo />').Save("$HOME/t.xml")
Get-Content -AsByteStream $HOME/t.xml | Select-Object -First 3
```

The above yields:

```none
239
187
191
```

which are the byte values of the UTF-8 BOM (`0xef 0xbb 0xbf`)

**Comment by krwq** [comment]
Per conversation in https://github.com/dotnet/dotnet-api-docs/issues/2377 I'm reopening this issue

**Comment by mklement0** [comment]
Taking a look at: 

https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs#L1372-L1390

The behavior comes down to `System.Text.Encoding.GetEncoding("utf-8")` (or case variants therefore) returning an UTF-8 encoding instance _with BOM_, just like `System.Text.Encoding.UTF8` does.

Ideally, `System.Text.Encoding.GetEncoding("utf-8")` would return a _BOM-less_ encoding, which would implicitly fix the problem at hand.

I'm mentioning this on the off chance that such a change would be considered - it certainly has the potential to break existing code.

Otherwise, the plan is to simply do:

```csharp
return String.Equals(value, "utf-8", StringComparison.OrdinalIgnoreCase) ? 
            new System.Text.UTF8Encoding(encoderShouldEmitUTF8Identifier: false) : 
            System.Text.Encoding.GetEncoding(value);
```

**Comment by krwq** [comment]
@mklement0 I'd move the "utf-8" check to the inner `if` and that's a fine fix as long as:

```csharp
    XmlDocument doc = ...; // construct your doc here
    XmlWriterSettings settings = new XmlWriterSettings();
    settings.Encoding = Encoding.UTF8;
    XmlWriter writer = XmlWriter.Create(stream, settings);
    doc.Save(writer);
```

will cause the BOM to show up.

Please make sure to add test case with both of those when creating a fix

**Comment by sharwell** [comment]
> The most portable way to write text to a file is to add BOM.

I agree with this. Many times I've had to fix documents that ended up with replacement characters due to the lack of a BOM. I've never had to fix replacement characters in a document that was written with a BOM.

The W3C guidance for BOM was written with the assumption that all systems are defaulting to Unicode encodings. When this is true, the BOM does not help. However, this is _not_ the case worldwide, and the BOM provides the best available approach to avoiding misinterpretation of file contents. If we have an API that historically wrote a BOM, we should preserve that API in that form.

**Comment by mklement0** [comment]
@sharwell, the default behavior of `.Save()` in the _absence_ of a `charset` attribute - _BOM-less_ UTF-8 - contradicts your premise. As argued before, there is no justification for modifying this behavior based on the _incidental_ aspect of making that _default_ behavior _explicit_ via `charset="UTF-8"`. Any compliant XML parser should assume UTF-8 _by default_ and, certainly, also if `charset="UTF-8"` is present - there is _no_ need for a BOM.

**Comment by krwq** [comment]
@sharwell I initially felt similar push back and I've also had to deal with weird encoding issues in the past but at the same time:
- I do not expect any reasonable XML parser to break on a UTF-8 document with explicit charset set to UTF-8
- The proposed behavior feels more consistent
- It does in fact make integrating with Unix tools better
- Suggested solution has a workaround if someone needs old behavior back

on the other hand there is:
- possibility of people porting older code to newer version of the framework which could possibly break on some odd XML parser

Having said that I do not feel super strongly about fixing this issue but it feels better to me to fix it than to document the inconsistency as normal behavior.

**Comment by sharwell** [comment]
> I do not expect any reasonable XML parser to break on a UTF-8 document with explicit charset set to UTF-8

Note that my concerns are generally not related to XML parsers. They are related to text editors used for opening XML files for editing.

**Comment by mklement0** [comment]
@sharwell, text editors that do _not_ default to UTF-8 are going the way of the dodo, for good reasons.

**Comment by krwq** [comment]
@sharwell I've personally have seen more people having issues with BOM than lack of it because they didn't understand it. IMO there is currently more tools/editors which will default to UTF-8 than than tools which default to different encoding than UTF-8 and will have issues displaying it but I do not have specific metrics. Unix tools are good example - they all assume UTF-8 without BOM.

Looking at the stats: https://w3techs.com/technologies/history_overview/character_encoding/ms/y it made more sense to not fix this issue couple years ago than now. IMO we should try to unify encoding on the internet and UTF-8 seems most dominant at the moment. If any tool can't open it or doesn't default to UTF-8 it's probably a misdesign as per stats it won't be able to open majority of the documents on the web.

**Comment by mklement0** [comment]
@sharwell, to add to @krwq's comments:

Concern about Unix tools choking on a UTF-8 BOM was indeed what motivated me to create this issue in the first place.

As for your text-editor concerns: I've looked at the behavior of popular editors available on Windows with respect to BOM-less UTF-8:

Atom, Sublime Text 3,  Visual Studio Code, Notepad++, Notepad:
* Correctly read BOM-less UTF-8-encoded files.
* Create BOM-less UTF-8 files by default (for Notepad, this was different up to at least Windows 7; not sure about Windows 8.1).

(When reading a BOM-less file:
* Atom and Visual Studio Code blindly assume UTF-8 and therefore potentially misread files.
* Notepad, Notepad++, and Sublime Text 3 fall back to ANSI encoding, presumably on encountering invalid-as-UTF-8 bytes in the input.)

**Comment by buyaa-n** [comment]
> /cc @buyaa-n my recommendation is to not change the behavior of this API

I was convinced to updated this when we were [triaging about this in the doc repo](https://github.com/dotnet/dotnet-api-docs/issues/2377#issuecomment-789952254), with the above comments, i convinced even more that we need this 😄 

tagging @GrabYourPitchforks if he has any concerns

**Comment by GrabYourPitchforks** [comment]
I agree with the comment at https://github.com/dotnet/runtime/issues/28218#issuecomment-791742745. The default `Encoding.UTF8` singleton should be no-BOM, which would the behavior of creating a fresh `new UTF8Encoding()` instance. The BOM is a crutch that may have had a purpose 20 years ago when .NET was first introduced, but nowadays it doesn't serve a legitimate purpose.

Note: _readers_ may still want to honor BOM if present. Changing `Encoding.UTF8` to be no-BOM by default should only affect _writers_, telling them not to emit the BOM when generating new documents.

Of course, changing `Encoding.UTF8` to be no-BOM may ripple throughout the .NET ecosystem, so a change of that magnitude would need to go through some kind of compat review. But I'd be strongly supportive of it under "standards-compliant" pillar that we have set for ourselves.

**Comment by sksk571** [comment]
When switching from .NET5 to .NET6 the following code starts emitting BOM together with encoding in XML declaration. I assume this happens because XmlSerializer defaults to Encoding.UTF8 which has BOM.

```
using System.Text;
using System.Xml.Serialization;

var obj = new object();
XmlSerializer serializer = new XmlSerializer(obj.GetType());

var stream = new MemoryStream();
serializer.Serialize(stream, obj);
```

**Comment by eOkadas** [comment]
Hi, when switching to NET 6 (from net 5 and netcore 3.1) we get the BOM for UTF8 added to our XML outputs.

The previous workaround offered by @sksk571 didn't work. 
[see fiddle with the workaround working in 4.7.2](https://dotnetfiddle.net/A3GYfv)
```XML
<?xml version="1.0"?>
<Test xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" />
```
  [see fiddle with the workaround NOT working in Net6]( https://dotnetfiddle.net/g1hS06)

```XML
﻿<?xml version="1.0" encoding="utf-8"?><Test xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" />
```

**Comment by jedediah617** [comment]
See the reply here https://stackdev.space/questions/loadingxmldocumentloadxml-and-the-xml-declarations-encoding-attribute
